### ISSUE TITLE
When are missing Q-thingey references fatal?

### ISSUE THREAD

**Issue by MikeBishop** [quicwg/base-drafts#1176]
@afrind raises a good hypothetical case:

- Encoder sends a header block on a request stream, which is delayed in the network
- Encoder side resets the stream, which is lost in the network
- Encoder sends another header block on the control stream that evicts an entry referenced in the first block, and this is delivered immediately.
- The first delayed header block arrives

From the encoder's side, it's valid to evict something referenced by the first block, because the stream has been reset.  From the decoder's side, there's a reference to something already evicted, because it doesn't know about the stream reset.

The solution to this seems to be that missing references on request/push streams are stream errors, not connection errors.  Missing references on the control stream still need to be connection errors.

**Comment by afrind** [comment]
I think the scenario that leads to this kind of error will be extremely rare, but I dislike the idea of there being some background level of 'expected' decoding errors.  It will make it harder to debug/investigate buggy implementations.  In the case where the missing reference came from an encoder that is broken or otherwise out of sync with the decoder, worse things could happen from continuing to decode subsequent requests.

The only other way I can think of to ensure such a scenario doesn't inadvertently kill a good connection is to have the encoder clear the outstanding references of the reset stream only after it's received the ACK for that RST_STEAM.  I'm already not crazy about having transport identifiers tied into header encoding state, and having to further couple to transport acknowledgements would be even worse.  Still, a careful encoder could take this approach.

**Comment by krasic** [comment]
+1 that moving to stream errors might introduce greater harm than they solve.   

This is similar to the more general reset issue we hoped to avoid moving updates out of the request-response stream.  

As in that discussion, another option is to have the encoder refrain from resetting streams that have outstanding headers.   I.e. an HQ mapping only solution is possible here:

the encoder side can have a notion of stream going away (don't send new data, drop any incoming data, don't start any new headers), but wait for header acks before actually sending reset.   To stay conservative, it could use a timer to trigger sending the reset anyway if the expected ack takes too long.  [ of course if we went this far, one might argue the whole update on separate stream approach is no longer needed. ]

Alternatively, if the occurrences of these are rare enough, perhaps status quo of letting such connections close with error is acceptable.   A server could track how many streams are reset with headers outstanding to help guage whether the number of missing header connection close errors are reasonable.

**Comment by MikeBishop** [comment]
#1314 includes the stream error, but it has the drawbacks mentioned here.  A better option that fits within the current model might be this:  Rather than the encoder counting all blocks as out-of-scope after it resets a stream, the decoder sends an acknowledgment either when it reads a header block from a stream or when it discards all remaining data from a stream.

This would require a second type of acknowledgement that is triggered by a reset stream, which is a little icky.

**Comment by dtikhonov** [comment]
+1 to @afrind for catching this.

Perhaps the HoL minimization/stream reset interplay is indeed a general issue, as @krasic points out, and is never going away.  In that case, we have to deal with it.

I think that having the decoder return a stream error is not so bad:
- If the encoder knows that the stream was reset, it can ignore this error;
  - This adds a requirement that the application (HQ layer) notify the encoder that it reset the stream.
- Otherwise, the encoder can escalate (promote?) this to a connection error and close the connection.

Introducing the concept of allowed stream errors on reset streams is better than dipping into the transport layer or using timers.
